src/devices/machine/cs8900a.cpp

```diff
@@ -0,0 +1,1522 @@
+#define CS8900_PP_ADDR_SE_BUFEVENT     0x012C /*   R- - 4.4.13., p. 59 - 000C */
+#define CS8900_PP_ADDR_SE_RXMISS       0x0130 /*   R- - 4.4.14., p. 60 - 0010 */
+#define CS8900_PP_ADDR_SE_TXCOL        0x0132 /*   R- - 4.4.15., p. 61 - 0012 */
+#define CS8900_PP_ADDR_SE_LINEST       0x0134 /*   R- - 4.4.17., p. 63 - 0014 */
+#define CS8900_PP_ADDR_SE_SELFST       0x0136 /*   R- - 4.4.19., p. 65 - 0016 */
+#define CS8900_PP_ADDR_SE_BUSST        0x0138 /* # R- - 4.4.21., p. 67 - 0018 */
+#define CS8900_PP_ADDR_SE_TDR          0x013C /*   R- - 4.4.23., p. 69 - 001C */
+
+#define MAX_TXLENGTH 1518
+#define MIN_TXLENGTH 4
+
+#define MAX_RXLENGTH 1518
+#define MIN_RXLENGTH 64
+
+#define CS8900_TX_IDLE 0
+#define CS8900_TX_GOT_CMD 1
+#define CS8900_TX_GOT_LEN 2
+#define CS8900_TX_READ_BUSST 3
+
+#define CS8900_RX_IDLE 0
+#define CS8900_RX_GOT_FRAME 1
+
+#ifdef CS8900_DEBUG_FRAMES
+    #define return( _x_ ) \
+    { \
+        int retval = _x_; \
+        \
+        log_message(cs8900_log, "%s correct_mac=%u, broadcast=%u, multicast=%u, hashed=%u, hash_index=%u", (retval? "+++ ACCEPTED":"--- rejected"), *pcorrect_mac, *pbroadcast, *pmulticast, *phashed, *phash_index); \
+        \
+        return retval; \
+    }
+#endif
+
+#define PP_PTR_AUTO_INCR_FLAG 0x8000 /* auto increment flag in package pointer */
+#define PP_PTR_FLAG_MASK      0xf000 /* is always : x y 1 1 (with x=auto incr) */
+#define PP_PTR_ADDR_MASK      0x0fff /* address portion of packet page pointer */
+
+
+#define LO_BYTE(x)      (u8)((x) & 0xff)
+#define HI_BYTE(x)      (u8)(((x) >> 8) & 0xff)
+#define LOHI_WORD(x,y)  ( (u16)(x) | ( ((u16)(y)) <<8 ) )
+
+//------------------- END #defines ---------------------
+
+
+unsigned long cs8900a_device::crc32_buf(const char *buffer, unsigned int len)
+{
+    unsigned long crc;
+    const char *p;
+
+    crc = 0xffffffff;
+    for (p = buffer; len > 0; ++p, --len)
+        crc = (crc >> 8) ^ crc32_table[(crc ^ *p) & 0xff];
+    
+    return ~crc;
+}
+
+void cs8900a_device::cs8900_set_tx_status(int ready,int error)
+{
+    u16 old_status = GET_PP_16(CS8900_PP_ADDR_SE_BUSST);
+    
+    /* mask out TxBidErr and Rdy4TxNOW */
+    u16 new_status = old_status & ~0x180;
+    if(ready)
+        new_status |= 0x100; /* set Rdy4TxNOW */
+    if(error)
+        new_status |= 0x080; /* set TxBidErr */
+    
+    if(new_status!=old_status) {
+        SET_PP_16(CS8900_PP_ADDR_SE_BUSST,new_status);
+#ifdef CS8900_DEBUG_RXTX_STATE
+        log_message(cs8900_log,"TX: set status Rdy4TxNOW=%d TxBidErr=%d",
+                    ready,error);
+#endif
+    }
+}
+
+void cs8900a_device::cs8900_set_receiver(int enabled)
+{
+    rx_enabled = enabled;
+    rx_state = CS8900_RX_IDLE;
+
+    rxevent_read_mask = 3; /* was L or H byte read in RXEVENT? */
+}
+
+void cs8900a_device::cs8900_set_transmitter(int enabled)
+{
+    tx_enabled = enabled;
+    tx_state = CS8900_TX_IDLE;
+    
+    cs8900_set_tx_status(0,0);
+}
+
+void cs8900a_device::device_reset(void)
```

Comment:
We generally don’t put `void` in the parentheses for functions that take no arguments, since C++ doesn’t support K&R-style functions.